include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,1723 @@
+        "base_abstract_domain_t and field_abstract_domain_t must be the same");
+  }
+
+  boost::optional<obj_id_t> get_obj_id(variable_t rgn) {
+    if (!m_flds_id_map) {
+      return boost::none;
+    }
+    auto it = (*m_flds_id_map).find(rgn);
+    if (it == (*m_flds_id_map).end()) {
+      return boost::none;
+    }
+    return it->second;
+  }
+
+  obj_id_t get_obj_id_or_fail(variable_t rgn) {
+    assert(m_flds_id_map);
+    auto it = (*m_flds_id_map).find(rgn);
+    if (it == (*m_flds_id_map).end()) {
+      CRAB_ERROR(domain_name(),
+                 "::get_odi, the odi map does not include the region ", rgn,
+                 " belonging to an abstract object");
+    }
+    return it->second;
+  }
+
+  void get_obj_flds(obj_id_t id, variable_vector_t &obj_flds) const {
+    assert(m_flds_id_map);
+    for (auto kv : (*m_flds_id_map)) {
+      if (kv.second == id) {
+        obj_flds.push_back(kv.first);
+      }
+    }
+  }
+
+  bool is_rgn_obj_id(variable_t rgn, obj_id_t id) { return id == rgn; }
+
+  void object_write(crab_os &o) const { // a special output for object domain
+    // not using api from seperate domain
+    if (m_odi_map.is_bottom()) {
+      o << "Object = _|_";
+    } else if (m_odi_map.is_top()) {
+      o << "Object = {}";
+    } else {
+      assert(m_flds_id_map != nullptr);
+      for (auto it = m_odi_map.begin(); it != m_odi_map.end();) {
+        o << "Object( ";
+        obj_id_t id = it->first;
+        variable_vector_t vars;
+        get_obj_flds(id, vars);
+        std::sort(vars.begin(), vars.end());
+        for (auto &v : vars)
+          o << v << " ";
+        o << ") = ";
+        auto dom = it->second;
+        dom.write(o);
+        ++it;
+        if (it != m_odi_map.end()) {
+          o << "; ";
+        }
+        o << "\n";
+      }
+    }
+  }
+
+  void move_singleton_to_odi_map(variable_t rep) {
+    variable_vector_t flds_vec;
+    get_obj_flds(rep, flds_vec);
+
+    base_abstract_domain_t tmp_base(m_base_dom);
+    tmp_base.project(flds_vec);
+
+    // Be careful of the following operation if type of base dom and object dom
+    // are different
+    ERROR_IF_DOM_TYPE_NOT_SAME();
+
+    field_abstract_domain_t res_obj_dom;
+    const field_abstract_domain_t *obj_dom_ref = m_odi_map.find(rep);
+    // In this case, the odi map must not exist that object domain
+    assert(!obj_dom_ref);
+    res_obj_dom = tmp_base;
+
+    m_odi_map.set(rep, res_obj_dom);
+    m_base_dom.forget(flds_vec);
+  }
+
+  void join_singleton_with_non_singleton(variable_t rep,
+                                         const base_abstract_domain_t &base_dom,
+                                         odi_map_t &odi_map) const {
+    variable_vector_t flds_vec;
+    get_obj_flds(rep, flds_vec);
+
+    base_abstract_domain_t tmp_base(base_dom);
+    tmp_base.project(flds_vec);
+
+    // Be careful of the following operation if type of base dom and object dom
+    // are different
```

Comment:
This is something we can check with `static_assert` at the beginning of the class, right?

Reply:
I used to have it in a function:
https://github.com/LinerSu/crab/blob/5018d40d7b7cca55ad05225e0d2af848f3d42125/include/crab/domains/object_domain.hpp#L547-L552
Now I see the `static_assert` can be a member declaration. I will change that.